### ISSUE TITLE
[Serve] Marlin FP8 wide path overruns c_tmp: buffer sized for marlin_fixed_m(), run at prob_m = T

### ISSUE THREAD

**Issue by flaviusburca** [invergent-ai/surogate#111]
**Severity:** high &nbsp;·&nbsp; **Location:** [`csrc/src/serve/ops/linear/marlin/marlin_plane.cu:561`](https://github.com/invergent-ai/surogate/blob/7f12387624e04c498dbd8df879eaeff13dea0b25/csrc/src/serve/ops/linear/marlin/marlin_plane.cu#L561) &nbsp;·&nbsp; **Audited at:** `7f123876`

### Defect
The shared Marlin fp32 reduce buffer is allocated once, sized from `marlin_c_tmp_floats(sms, marlin_fixed_m())`, i.e. `sms * min(round16(fixed_m), 64) * max_thread_n` floats. Every other Marlin call site clamps prob_m to `marlin_fixed_m()` (the padded band), but `marlin_fp8_run`'s wide branch passes the caller's raw `t` as prob_m. Inside the kernel the per-CTA slice of C_tmp is `c_size = (thread_m_blocks*16) * (thread_n_blocks*16) * sizeof(float)/16` int4 at offset `locks_off * c_size`, with locks_off reaching gridDim.x == sms. With fixed_m == 32 the host reserves 32*256 floats per locks_off slot while the kernel writes 64*256 when thread_m_blocks reaches 4 and the large-batch config picks thread_n = 256 — exactly 2x the reservation. The overrun lands in the `lock_buf` allocated immediately after c_tmp_buf in ensure_scratch, then in whatever follows.

### Failure scenario
Engine started with SUROGATE_SERVE_MARLIN_FP8=1 and the default max_concurrency (1) so marlin_set_fixed_m(1) -> g_fixed_m = 32; ensure_scratch allocates c_tmp = marlin_c_tmp_floats(170, 32) = 170*32*256 = 1,392,640 floats (5.57 MB). The 27B's FP8 down projection (n=5120, k=17408) is adopted to MarlinTiles by marlin_fp8_maybe_adopt, so linear_add.cpp takes the adopted branch and calls marlin_fp8_run with a 512-token prefill. marlin_mm computes thread_m_blocks = min(ceil(512/16), 4) = 4 and determine_exec_config selects large_batch_thread_configs[0] = {thread_k 64, thread_n 256, 256 threads} (5120 % 256 == 0, 17408 % 64 == 0, and the generated selector has that instantiation). c_size becomes 64*256 = 16,384 floats per locks_off, and global_reduce_fp32 writes up to (170+1)*16,384 = 2.8M floats = 11.2 MB into a 5.57 MB allocation — ~5.6 MB of device memory past the end, starting with the Marlin lock array, producing corrupted cross-CTA barrier flags (hangs or garbage logits) and heap corruption in whatever else follows.

### Evidence
````cpp
marlin_plane.cu:113  `const std::size_t c_tmp_bytes = marlin_c_tmp_floats(sms, marlin_fixed_m()) * sizeof(float);`
marlin_plane.cu:553-563
```
    const bool adopted = weight.layout == QuantLayout::MarlinTiles;
    if (adopted || t > marlin_fixed_m()) {
        ...
        const MarlinScratch wide_scratch = marlin_scratch();
        if (wide_scratch.c_tmp == nullptr || wide_scratch.locks == nullptr) { return false; }
        marlin_gemm_bf16(x.data, wide.b_packed, wide.scales, out.data, wide_scratch.c_tmp,
                         wide_scratch.locks, t, weight.n, weight.k, /*group_size=*/-1,
                         /*b_is_fp8=*/true, wide_scratch.sm_count, stream);
```
marlin_gemm.cu:445-449
```
std::size_t marlin_c_tmp_floats(int sm_count, int max_tokens) {
    const int m16 = ((max_tokens + 15) / 16) * 16;
    const int max_m = m16 < 64 ? m16 : 64;
    return static_cast<std::size_t>(sm_count) * max_m * MARLIN_NAMESPACE_NAME::max_thread_n;
}
```
vendor/marlin_template.h:1598-1609
```
    constexpr int tb_m = thread_m_blocks * 16;
    constexpr int tb_n = thread_n_blocks * 16;
    constexpr int c_size = tb_m * tb_n * sizeof(float) / 16;
    ...
    int c_cur_offset = locks_off * c_size;
```
````

### Suggested fix
Size c_tmp for the largest prob_m the route can ever run, not the decode band: allocate `marlin_c_tmp_floats(sms, 64)` unconditionally (64 is the vendor's ceiling, `max_thread_m_blocks * 16`), or grow it in ensure_scratch alongside gemm_out/a_pad and have marlin_fp8_run's wide branch decline when the current c_tmp is smaller than `sm_count * 64 * max_thread_n` floats.

### Verification
Found in an audit of `csrc/src/serve`, then upheld by two independent verifiers: one tried to refute it, the other traced a concrete path to it.
Corrections the verifiers made to the original report:

> The defect is real, but some details are off. (1) The overrun is smaller than claimed. For the 27B down projection at t=512 on a 5090, the kernel sees 160 mn-tiles, which is fewer than the 170 SMs. That makes locks_off equal to the column index, so its maximum is 159, not 170. The kernel touches 160 x 16,384 floats (10.49 MB) against 5.57 MB allocated. Slots 85-159 are out of bounds, about 4.9 MB. (2) Nothing guarantees that lock_buf sits right after c_tmp_buf. They are separate cudaMalloc calls, so the write could corrupt any neighbouring device allocation or fault with an illegal address. (3) The path only runs with SUROGATE_SERVE_MARLIN_FP8=1, which is off by default (marlin_plane.cu:487-491). The undersizing happens whenever fixed_m==32, which covers any max_concurrency<=32 and SUROGATE_SERVE_MARLIN_NARROW=1. With the opt-in set, every adopted FP8 prefill wider than 32 tokens can overrun, so the severity is high rather than critical.

> The defect is real, with three corrections. First, the write does not reach (170+1)*16384 floats for this shape. locks_off is bounded by the number of column tiles (8*20 = 160 < sms), so the reduce writes up to 160*16384 = 2.62M floats (10.5 MB) into the 5.57 MB buffer. Second, not every t above fixed_m overruns. It needs (column tiles)*tb_m*tb_n > sms*32*256. For the 5120x17408 down projection that means t of about 320 or more. For the fused gate_up (n=34816, 136 column tiles) it happens at any t of 49 or more. t=40 on the down projection stays in bounds (20 slots of 12288 floats). Third, nothing guarantees the overrun lands in `lock_buf`: the buffers come from separate `cudaMalloc` calls, so it corrupts whatever device memory happens to follow. The trigger is `SUROGATE_SERVE_MARLIN_FP8=1` (off by default, and documented as a measured net loss) with `max_concurrency` of 32 or less, or with `SUROGATE_SERVE_MARLIN_NARROW=1`. Because it is opt-in, high fits better than critical. Fix: size `c_tmp` with `marlin_c_tmp_floats(sms, 64)`, or check the buffer size before the wide call.

_Line numbers quoted by the verifiers may refer to a later tree; the permalinks above are pinned to `7f123876`._

**Comment by flaviusburca** [comment]
Fixed in commit `40e367fa5c84d6a5a6d6108f8022bcfcef53f637`. The fix is committed locally and validated; publication to the remote repository is pending.

Marlin reduction scratch is now sized for the maximum FP8 tile width used by the wide path.

Validation: The reported wide-shape regression passes at widths 1/32/64/129/512 in eager and CUDA graph execution. CUDA memcheck reports no errors.
